### ISSUE TITLE
'weight' column in balance matrix

### ISSUE THREAD

**Issue by dmalzl** [deeptools/HiCExplorer#538]
Hi,

I am currently trying to understand the mechanisms behind the matrix balancing algorithms a bit more in-depth which raised a question I couldn't really find an answer to. I am using the KR algorithm to generate a genome-wide balanced matrix and the resulting file does not only contain the matrix and its balanced values but also a column named `'weight'`. I was just wondering what this additional values are meant for. Would be great it if you could enlighten me.

Thank you for your time in advance.

**Comment by joachimwolff** [comment]
Hi,

please have a look at the cooler file format definition. https://academic.oup.com/bioinformatics/article-abstract/36/1/311/5530598?redirectedFrom=fulltext

'weight' contains the correction factors calculated by KR i.e. the format stores the raw (not corrected) values and the correction factors. At the time the data is loaded, the correction factors are applied.

Best,

Joachim

**Comment by dmalzl** [comment]
Thanks for the quick reply. This I already noticed, but what I don't get is, if the format stores the raw not corrected values, why is it that the file I get from hicCorrectMatrix already contains float values for each respective entries. Doesn't that mean that these values already experienced some kind of correction since the raw matrix definitely contains integer values (as expected from the ditag counting).

Best,
Daniel

**Comment by joachimwolff** [comment]
We apply the correction factors at the time the matrix is loaded which leads to float values. At the time the matrix is stored these correction factors are removed by the inverse mathematical operation and leads to the 'raw' value again. The raw value is now stored as float.

**Comment by dmalzl** [comment]
Okay, I see. However, when I load the corrected H5 using the hicmatrix library I get very different results from the raw matrix (see screenshot attached). I had a look at the code of the hicmatrix library but could not find a line where correction factors are explicitly applied. So this must be taking place somewhere in hicCorrectMatrix and I suspect line 737 through 741 as the source of this where it seems that you replace the values in the matrix with the corrected matrix values if the output format is h5:
```
            if args.outFileName.endswith('.h5'):
                corrected_matrix = kr.get_normalised_matrix(True)

    if args.outFileName.endswith('.h5'):
        ma.setMatrixValues(corrected_matrix)
    # if
    ma.setCorrectionFactors(correction_factors)
```

so to me it looks like the corrected values are stored in the matrix rather than the raw values.
I just ask because when I am converting to cooler with hicConvertFormat, I was a little bit confused what to do whether to load the matrix balanced (applying the correction factors) or leave it as it is. Am I now right that I have to load the converted matrix as it is without applying the correction factors or am I missing something burried in the kr_balancing module.

![image](https://user-images.githubusercontent.com/52696584/81932319-e9c3d500-95eb-11ea-8b12-90c39fab7a76.png)

**Comment by joachimwolff** [comment]
What I wrote applies only for the cool format, not for h5. In h5 the corrected values are stored and no raw values are present.

**Comment by joachimwolff** [comment]
I also want to point out that you can use this lib as you like because it is open source. However, we do not see it as a public API and can change the behaviour and parameter options any time without much communication if it is necessary for the two main tools of us: HiCExplorer and pyGenomeTracks.

**Comment by dmalzl** [comment]
Ah, I get it. I was just wondering because I first used h5 and then converted to cooler in order to use the data with cooltools and then nothing made sense because the correction factors were applied to the already corrected values. Anyway, I see that's why I copied the relevant parts for my tools, I just used it for this quick and dirty example. 

Would be nice if somewhen in the future the discrepancy between h5 and cool in hicexplorer is mentioned more explicitly then somewhere buried in the commandline arguments and code so people can immediately see that there is a difference between the two things in case they want to use it with some cooler related things. But that's just a suggestion for docs enhancement.

Anyway, thanks for the discussion.
Best, 
Daniel
